third_party/move/move-model/bytecode/src/astifier.rs

```diff
@@ -1359,152 +1371,128 @@ impl<'a> ExpRewriterFunctions for IfElseTransformer<'a> {
 impl<'a> IfElseTransformer<'a> {
     /// Attempts to create an if-then-else from the given expression.
     /// This recognizes the pattern below, as produced by the AST
-    /// generator:
+    /// generator. Here, with 'does not branch out of a loop' we mean
+    /// that an expressions does not contain any unbound break or
+    /// continue statements which point outside the given loop:
     ///
     /// ```move
     /// loop {
+    ///   loop { // no loop header
+    ///     ( if (c_i) break; )*
+    ///     <else-branch> // no reference to inner loop
+    ///     break[1]
+    ///   }
+    ///   <then-branch>
+    /// }
+    /// ==>
+    /// if (c1 || .. || cn) {
     ///   loop {
-    ///     if (c) break; // exclusive binding to inner loop
-    ///     <else-branch>
-    ///     break[1]; // 1st binding to outer loop
+    ///     <then-branch>
+    ///   }
+    /// else {
+    ///   loop {
+    ///     <else-branch> where loop_nest -= 1
     ///   }
-    ///   <then-branch> // must not reference outer loop
-    ///   break | abort | return; // 2nd binding to outer loop or exit
     /// }
     /// ```
-    ///
-    /// The comment at the break indicates that both loops
-    /// must have exactly one associated break statement.
     fn try_make_if_else(&self, exp: Exp) -> Option<Exp> {
         let node_id = exp.node_id();
         let default_loc = self.builder.env().get_node_loc(node_id);
-        let (outer_id, outer_body) = self.builder.match_loop(exp)?;
-        let (outer_first, outer_rest) = self.builder.extract_first(outer_body);
-        let (mut outer_rest, outer_break) = self
-            .builder
-            .extract_last(self.builder.seq(&default_loc, outer_rest));
-
-        let mut loop_binding_count = self.loop_to_cont[&outer_id].len();
-        if outer_break.is_loop_cont(Some(0), false) {
-            // This is a break against the outer loop, count
-            loop_binding_count -= 1
-        } else if matches!(
-            outer_break.as_ref(),
-            ExpData::Call(_, Operation::Abort, ..) | ExpData::Return(..)
-        ) {
-            // Not counting for loop bindings, and put expression
-            // back
-            outer_rest.push(outer_break)
-        } else {
-            return None;
-        }
 
-        let (inner_id, inner_body) = self.builder.match_loop(outer_first)?;
-        if self.loop_to_cont[&inner_id].len() != 1 {
+        // Match the pattern as described above
+        let outer_body = match_ok!(exp.as_ref(), ExpData::Loop(_, _0))?;
```

Comment:
I think you can use standard Rust for this:
```
let ExpData::Loop(_, outer_body) = exp.as_ref() else { return None };
```
No need for fancy macros anymore.  https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html

Reply:
I disagree, these fancy macros are extremely helpful and make the code much more readable compared to those actually `let .. else`. @vineethk were discussing the desire for this also in the context of the linter which frequently needs to match expressions, those macros are the solution.